crud.go

```diff
@@ -152,6 +157,230 @@ func (cc *CrudComponent) GetReplica(ctx context.Context, opts *GetReplicaOptions
+		resp, err := client.GetReplica(ctx, &memdx.GetReplicaRequest{
+			CollectionID: collectionID,
+			Key:          opts.Key,
+			VbucketID:    vbID,
+			CrudRequestMeta: memdx.CrudRequestMeta{
+				OnBehalfOf: opts.OnBehalfOf,
+			},
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		value, datatype, err := cc.compression.Decompress(memdx.DatatypeFlag(resp.Datatype), resp.Value)
+		if err != nil {
+			return nil, err
+		}
+
+		return &GetAllReplicasResult{
+			Value:     value,
+			Flags:     resp.Flags,
+			Datatype:  datatype,
+			Cas:       resp.Cas,
+			IsReplica: true,
+		}, nil
+	}
+
+	var endpoints []string
+	var mu sync.Mutex
+	var sendLock sync.Mutex
+	var returnedResults uint32
+	var numReplicas atomic.Uint32
+
+	temp, err := cc.vbs.NumReplicas()
+	if err != nil {
+		return nil, err
+	}
+	numReplicas.Store(uint32(temp))
+
+	// This retry orchestrator handles request level retryable errors, errors which impact every replica request,
+	// e.g the collection ID not yet being consistent
+	_, err = OrchestrateMemdRetries(ctx, cc.retries, func() (any, error) {
+		return OrchestrateMemdCollectionID(ctx, cc.collections, opts.ScopeName, opts.CollectionName,
+			func(collectionID uint32, manifestID uint64) (any, error) {
+
+				// We are past the point of no return. From here on the request cannot error, e.g a nil error will always be
+				// returned from GetAllReplicas, however individual replica reads can push an error to the channel.
+				for i := uint32(0); i <= numReplicas.Load(); i++ {
+					go func(replicaID uint32) {
+						for {
+							res, err := OrchestrateReplicaRead(ctx, cc.vbs, cc.nmvHandler, cc.connManager, opts.Key, replicaID,
+								func(ep string, vbID uint16, client KvClient) (*GetAllReplicasResult, error) {
+									mu.Lock()
+									if slices.Contains(endpoints, ep) {
+										mu.Unlock()
+										return nil, nil
+									}
+									mu.Unlock()
+
+									var res *GetAllReplicasResult
+									var err error
+
+									if replicaID == 0 {
+										res, err = getFn(collectionID, vbID, client)
+									} else {
+										res, err = getReplicaFn(collectionID, vbID, client)
+									}
+
+									mu.Lock()
+									// If a rebalance occured while we were executing the function to get the doc
+									// a thread with a different replicaID may have been routed to the same endpoint,
+									// fetched the replica and added to the endpoints slice, so we repeat the check
+									if err == nil && !slices.Contains(endpoints, ep) {
+										endpoints = append(endpoints, ep)
+										mu.Unlock()
+										return res, err
+									}
+									mu.Unlock()
+									return nil, err
+								})
+
+							// The replicaID is being routed to a node from which we have already received a result
+							// therefore we wait and then try again, incase the replicaIds have been changed by a rebalance
+							if res == nil && err == nil {
+								time.Sleep(10 * time.Millisecond)
```

Comment:
Rather than a hardcoded wait here, shouldn't this be based on the retry manager in some way?

Reply:
Done!